java/ql/src/Telemetry/AutomodelEndpointCharacteristics.qll

```diff
@@ -0,0 +1,325 @@
+private import semmle.code.java.dataflow.internal.ModelExclusions as ModelExclusions
+import AutomodelSharedCharacteristics as SharedCharacteristics
+import AutomodelEndpointTypes as AutomodelEndpointTypes
+
+module CandidatesImpl implements SharedCharacteristics::CandidateSig {
+  class Endpoint = DataFlow::ParameterNode;
+
+  class EndpointType = AutomodelEndpointTypes::EndpointType;
+
+  predicate isNegative(AutomodelEndpointTypes::EndpointType t) {
+    t instanceof AutomodelEndpointTypes::NegativeSinkType
+  }
+
+  // Sanitizers are currently not modeled in MaD. TODO: check if this has large negative impact.
+  predicate isSanitizer(Endpoint e, EndpointType t) { none() }
+
+  string getLocationString(Endpoint e) { result = e.getLocation().toString() }
+
+  predicate isKnownLabel(string label, string humanReadableLabel, EndpointType type) {
+    label = "read-file" and
+    humanReadableLabel = "read file" and
+    type instanceof AutomodelEndpointTypes::TaintedPathSinkType
+    or
+    label = "create-file" and
+    humanReadableLabel = "create file" and
+    type instanceof AutomodelEndpointTypes::TaintedPathSinkType
+    or
+    label = "sql" and
+    humanReadableLabel = "mad modeled sql" and
+    type instanceof AutomodelEndpointTypes::SqlSinkType
+    or
+    label = "open-url" and
+    humanReadableLabel = "open url" and
+    type instanceof AutomodelEndpointTypes::RequestForgerySinkType
+    or
+    label = "jdbc-url" and
+    humanReadableLabel = "jdbc url" and
+    type instanceof AutomodelEndpointTypes::RequestForgerySinkType
+    or
+    label = "command-injection" and
+    humanReadableLabel = "command injection" and
+    type instanceof AutomodelEndpointTypes::CommandInjectionSinkType
+  }
+
+  predicate isSink(Endpoint e, string label) {
+    exists(
+      string package, string type, boolean subtypes, string name, string signature, string ext,
+      string input
+    |
+      sinkSpec(e, package, type, subtypes, name, signature, ext, input) and
+      ExternalFlow::sinkModel(package, type, subtypes, name, [signature, ""], ext, input, label, _)
+    )
+  }
+
+  predicate isNeutral(Endpoint e) {
+    exists(string package, string type, string name, string signature |
+      sinkSpec(e, package, type, _, name, signature, _, _) and
+      ExternalFlow::neutralModel(package, type, name, [signature, ""], _)
+    )
+  }
+
+  additional predicate sinkSpec(
+    Endpoint e, string package, string type, boolean subtypes, string name, string signature,
+    string ext, string input
+  ) {
+    package = e.getEnclosingCallable().getDeclaringType().getPackage().toString() and
+    type = e.getEnclosingCallable().getDeclaringType().getName() and
+    subtypes = false and
+    name = e.getEnclosingCallable().getName() and
+    signature = ExternalFlow::paramsString(e.getEnclosingCallable()) and
+    ext = "" and
+    exists(int paramIdx | e.isParameterOf(_, paramIdx) | input = "Argument[" + paramIdx + "]")
```

Comment:
This means we only support indexed parameters (i.e. not `this` qualifiers or return values), which is totally fine. Just comment to highlight that.

Reply:
The limitation is from the application mode branch; I've kept it as-is to make the behaviour as similar as possible. But you're completely right!